File: algoritm/BOJ/11725_find_parent_tree/11725_find_parent_tree.py
from collections import deque
# 각 노드의 부모를 구해서 출력

N = int(input())

graph = [[] for _ in range(N + 1)]
for _ in range(N - 1):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)


# DFS로 풀기
# visited = [0] * (N + 1)

# def dfs(start):
#     for i in graph[start]:
#         if visited[i] == 0:
#             visited[i] = start
#             dfs(i)

# dfs(1)

# for i in range(2, N + 1):
#     print(visited[i])


# BFS로 풀기
queue = deque()
queue.append(1)

ans = [0] * (N + 1)

def bfs():
    while queue:
        now = queue.popleft()
        for next in graph[now]:
            if ans[next] == 0:
                ans[next] = now
                queue.append(next)

bfs()

result = ans[2:]

for i in result:
    print(i)

File: algoritm/BOJ/11725_find_parent_tree/11725_find_parent_tree.cpp
#include <iostream>
#include <vector>
#include <queue>


using namespace std;

#define MAX 100001

int N;
int arr[MAX];
bool visited[MAX];
vector<int> graph[MAX];

// dfs
void dfs(int k)
{
  visited[k] = true;
  for (int i = 0; i < graph[k].size(); i++)
  {
    int next = graph[k][i];
    if (!visited[next])
    {
      arr[next] = k;
      dfs(next);
    }
  }
}


// bfs
void bfs() 
{
  queue<int> q;
  visited[1] = true;
  q.push(1);

  while (!q.empty())
  {
    int parent = q.front();
    q.pop();

    for (int i = 0; i < graph[parent].size(); i++)
    {
      int child = graph[parent][i];
      if (!visited[child])
      {
        arr[child] = parent;
        visited[child] = true;
        q.push(child);
      }
    }
  }
}


int main ()
{
  cin >> N;

  for (int i = 0; i < N; i++)
  {
    int x, y;
    cin >> x >> y;
    graph[x].push_back(y);
    graph[y].push_back(x);
  }


  // dfs로 풀기
  // dfs(1);

  // for (int i = 2; i <= N; i++)
  // {
  //   cout << arr[i] << "\n";
  // }

  // bfs로 풀기
  bfs();

  for (int i = 2; i <= N; i++)
  {
    cout << arr[i] << "\n";
  }

  return 0;
}